text/0027-corejam.md

```diff
@@ -0,0 +1,706 @@
+
+The `prune` function is used by the Relay-chain Block Author in order to ensure that all Work Packages which make it through the Join stage are non-conflicting and valid for the present Relay-chain state. All Work Packages are rechecked using the same procedure on-chain and the block is considered malformed (i.e. it panics) if the result is not an empty `Vec`.
+
+The `prune` function has immutable access to the Work Class's child trie state, as well as regular read-only storage access to the Relay-chain's wider state.
+
+```rust
+fn get_work_storage(key: &[u8]) -> Result<Vec<u8>>;
+fn get_work_storage_len(key: &[u8]);
+```
+
+The amount of weight `prune` is allowed to use is a fixed multiple of the number of 
+
+#### Accumulate
+
+The second stage is that of Accumulate. The function signature to the `accumulate` entry-point in the Work Class's code blob is:
+
+```rust
+fn accumulate(results: Vec<(Authorization, Vec<(ItemHash, WorkResult)>)>);
+type ItemHash = [u8; 32];
+```
+
+As stated, there is an amount of weight which it is allowed to use before being forcibly terminated and any non-committed state changes lost. The lowest amount of weight provided to `accumulate` is defined as the number of `WorkResult` values passed in `results` to `accumulate` multiplied by the `accumulate` field of the Work Class's weight requirements.
+
+However, the actual amount of weight may be substantially more. Each Work Package is allotted a specific amount of weight for all on-chain activity (`weight_per_package` above) and has a weight liability defined by the weight requirements of all Work Items it contains (`total_weight_requirement` above). Any weight remaining after the liability (i.e. `weight_per_package - total_weight_requirement`) may be apportioned to the Work Classes of Items within the Report on a pro-rata basis according to the amount of weight they utilized during `refine`. Any weight unutilized by classes within one package may be carried over to the next package and utilized there.
+
+Work Items are identified by their hash (`ItemHash`). We provide both the authorization of the package and the item identifers and their results in order to allow the `refine` logic to take appropriate action in the case that an invalid Work Item was issued.
+
+_(Note for later: We may wish to provide a more light-client friendly Work Item identifier than a simple hash; perhaps a Merkle root of equal-size segments.)_
+
+```rust
+fn get_work_storage(key: &[u8]) -> Result<Vec<u8>>;
+fn get_work_storage_len(key: &[u8]);
+fn checkpoint() -> Weight;
+fn weight_remaining() -> Weight;
+fn set_work_storage(key: &[u8], value: &[u8]) -> Result<(), ()>;
+fn remove_work_storage(key: &[u8]);
+```
+
+Read-access to the entire Relay-chain state is allowed. No direct write access may be provided since `refine` is untrusted code. `set_storage` may fail if an insufficient deposit is held under the Work Class's account.
```

Comment:
I though this was already tricky after Asynchronous Backing. It should be enough to have access to the relay chain state root, from which storage proofs can be made. Because there is no guarantee, in the case of disputes especially, that a validator validating this package will have the right relay chain state.

Reply:
Nothing is disputable from `refine`. It's on-chain (executed by all validators).

Reply:
Sorry - that should be `accumulate`, not `refine`. It's not a problem since the code runs on-chain (i.e. checked by all validators at the time).

Reply:
If we can dispute these, that is not enough. Because it may have bern run on a fork we are not aware of.

Reply:
It is true that in the past we have designed the system to decouple execution of candidates from being familiar with the fork of the relay chain those candidates are anchored to. As Al mentions, this property improves the security of the disputes system, in that it is possible for validators to receive work packages to validate as (almost) completely self-contained items. This allows validators to participate in disputes even when relay chain block propagation is lagging behind. 

The "(almost)" there does some heavy lifting, particularly with respect to the validation code that is used to execute the work package. The storage pallet provides the same guarantees we currently expect for validation code, but the rest of the relay chain state is much more volatile.

Reply:
Ok, here in accumulate, it is fine.